src/platform/datapath_kqueue.c

```diff
@@ -721,126 +724,208 @@ CxPlatSocketContextInitialize(
     }
 
     //
-    // Set non blocking mode
+    // Set dual (IPv4 & IPv6) socket mode unless we operate in pure IPv4 mode
     //
-    Flags =
-        fcntl(
+    if (!ForceIpv4) {
+        Option = FALSE;
+        Result =
+            setsockopt(
+                SocketContext->SocketFd,
+                IPPROTO_IPV6,
+                IPV6_V6ONLY,
+                (const void*)&Option,
+                sizeof(Option));
+        if (Result == SOCKET_ERROR) {
+            Status = errno;
+            QuicTraceEvent(
+                DatapathErrorStatus,
+                "[data][%p] ERROR, %u, %s.",
+                Binding,
+                Status,
+                "setsockopt(IPV6_V6ONLY) failed");
+            goto Exit;
+        }
+    }
+
+    //
+    // The port may be shared across processors.
+    // Even if not, this is probably cheap.
+    //
+    Option = TRUE;
+    Result =
+        setsockopt(
             SocketContext->SocketFd,
-            F_GETFL,
-            NULL);
-    if (Flags < 0) {
+            SOL_SOCKET,
+            SO_REUSEADDR,
+            (const void*)&Option,
+            sizeof(Option));
+    if (Result == SOCKET_ERROR) {
         Status = errno;
         QuicTraceEvent(
             DatapathErrorStatus,
             "[data][%p] ERROR, %u, %s.",
             Binding,
             Status,
-            "fcntl(F_GETFL) failed");
+            "setsockopt(SO_REUSEADDR) failed");
         goto Exit;
     }
 
-    Flags |= O_NONBLOCK;
+
+    //
+    // bind() to local port if we need to. This is not necessary if we call connect afterward and there is no ask for particular
+    // source address or port. connect() will resolve that together in single system call.
+    if (!RemoteAddress || Binding->LocalAddress.Ipv6.sin6_port || !QuicAddrIsWildCard(&Binding->LocalAddress)) {
+        CxPlatCopyMemory(&MappedAddress, &Binding->LocalAddress, sizeof(MappedAddress));
+        if (MappedAddress.Ipv6.sin6_family == QUIC_ADDRESS_FAMILY_INET6) {
+            MappedAddress.Ipv6.sin6_family = AF_INET6;
+        }
+
+        if (ForceIpv4) {
+            MappedAddress.Ipv4.sin_family = AF_INET;
+            MappedAddress.Ipv4.sin_port = Binding->LocalAddress.Ipv4.sin_port;
+            // assume wildcard address for now!
+            CXPLAT_DBG_ASSERT(QuicAddrIsWildCard(&Binding->LocalAddress));
```

Comment:
Why do we have to assume wildcard?

Reply:
I did not see any test doing otherwise. So this reflects testing I did so far. 
If it is not wildcard we should probably assert that the address is either IPv4 or mapped IPv4.
I can fix that in next update.

Reply:
updated. While there is no test case, I added verification that if LocalAddress is not wildcard, it is IPv4 in some form.